CMakeLists.txt

```diff
@@ -2,156 +2,243 @@
 #
 # Copyright (c) 2015 Wenzel Jakob <wenzel@inf.ethz.ch>
 #
-# All rights reserved. Use of this source code is governed by a
-# BSD-style license that can be found in the LICENSE file.
+# All rights reserved. Use of this source code is governed by a BSD-style license that can be found
+# in the LICENSE file.
 
-cmake_minimum_required(VERSION 2.8.12)
+cmake_minimum_required(VERSION 3.7)
 
-if (POLICY CMP0048)
-  # cmake warns if loaded from a min-3.0-required parent dir, so silence the warning:
-  cmake_policy(SET CMP0048 NEW)
-endif()
-
-# CMake versions < 3.4.0 do not support try_compile/pthread checks without C as active language.
-if(CMAKE_VERSION VERSION_LESS 3.4.0)
-  project(pybind11)
+# VERSION 3.7...3.18, but some versions of VS have a patched CMake 3.11 that do not work properly
```

Comment:
This is neat way to deal with all of the policy warnings, I'll remember that for my next CMake project! It wasn't clear to me when reading these lines what the VS-specific issue is, and how those lines address it. Can you clarify?

Reply:
I've written about it here: https://cliutils.gitlab.io/modern-cmake/chapters/basics.html (I realize I need to update the text, I talk about the "upcoming release" of 3.12). The [bug report and resolution is here](https://developercommunity.visualstudio.com/content/problem/294070/cmake-312-version-requirements-definition-fails-fo.html). Do we only support the latest release of Windows compilers (IE, is it okay if an earlier release has an issue in the same family, like 15.3 vs. 15.4)? If that's the case, we can simplify this slightly. I have no idea how widespread this still is.

TL;DR version:

In CMake 3.12, they added the `VERSION 3.7...3.18` syntax - from CMake 3.12 on, this automatically sets the policies to the highest possible value, up to the maximum listed. By careful choice of syntax, it happens to be completely valid syntax (though you don't get the nice policy bump) in CMake < 3.12 on all official CMake versions! The problem is MSVC had a built-in fork of CMake < 3.12 that didn't support this.

Reply:
That's a really useful reference, thanks for sharing.